### PR TITLE
fix: Fix CBCS support in some platforms

### PR DESCRIPTION
**Pull request by avelad**
Fixes https://github.com/shaka-project/eme-encryption-scheme-polyfill/issues/62

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -600,6 +598,30 @@ function hasEncryptionScheme(mediaKeySystemAccess) {
...
+ * @param {?string} supportedScheme A guess at the encryption scheme this
+ *   supports.
+ * @return {boolean} True if the scheme is compatible.
+ */
+function checkSupportedScheme(scheme, supportedScheme) {
+  if (scheme && scheme != supportedScheme) {
+    if (scheme == 'cbcs') {
+      // Firefox supports CBCS since version 100, but does not yet implement
+      // encryptionScheme support.
+      if (parseInt(navigator.userAgent.split('Firefox/').pop(), 10) >= 100) {
```
Comment (joeyparrish) [inline index.js:612]: I would like to see this pulled out to a place where it is run once and cached/memoized.

File: index.js [hunk 2]
```diff
@@ -600,6 +598,30 @@ function hasEncryptionScheme(mediaKeySystemAccess) {
...
 }
 
+/**
+ * @param {?string} scheme Encryption scheme to check
+ * @param {?string} supportedScheme A guess at the encryption scheme this
+ *   supports.
+ * @return {boolean} True if the scheme is compatible.
+ */
+function checkSupportedScheme(scheme, supportedScheme) {
+  if (scheme && scheme != supportedScheme) {
```
Comment (joeyparrish) [inline index.js:608]: I would find this logic more readable if we flattened it and had early-returns for each special case, one at a time.  For example:

```js
if (!scheme) {
  // Not encrypted = always supported
  return true;
}

if (scheme == supportedScheme) {
  // The assumed-supported legacy scheme for this platform.
  return true;
}

if (scheme == 'cbcs' || scheme == 'cbcs-1-9') {
  if (isRecentFirefox || isChromecast) {
    // Firefox >= 100 supports CBCS, but doesn't support queries yet.
    // Older Chromecast devices are assumed to support CBCS as well.
    return true;
  }
}

return false;
```

File: index.js [hunk 3]
```diff
@@ -600,6 +598,30 @@ function hasEncryptionScheme(mediaKeySystemAccess) {
...
+    if (scheme == 'cbcs') {
+      // Firefox supports CBCS since version 100, but does not yet implement
+      // encryptionScheme support.
+      if (parseInt(navigator.userAgent.split('Firefox/').pop(), 10) >= 100) {
+        return true;
+      }
+      // All Chromecast supports CBCS.
+      if (navigator.userAgent.includes('CrKey')) {
```
Comment (joeyparrish) [inline index.js:616]: This should also be cached instead of parsing every time.

**Review by joeyparrish — CHANGES_REQUESTED** [review]

File: index.js (hunk 1 above)
Comment (avelad) [reply index.js]: Done!

File: index.js (hunk 2 above)
Comment (avelad) [reply index.js]: Done!

File: index.js (hunk 3 above)
Comment (avelad) [reply index.js]: Done!

**Comment by avelad** [comment]
@joeyparrish can you review it again? Thanks!

File: index.js [hunk 4]
```diff
@@ -369,10 +367,10 @@ class McEncryptionSchemePolyfill {
         configuration: requestedConfiguration,
       };
 
-      if (audioScheme && audioScheme != supportedScheme) {
+      if (audioScheme && !checkSupportedScheme(audioScheme, supportedScheme)) {
```
Comment (joeyparrish) [inline index.js:370]: You can drop the null check here, since checkSupportedScheme has it.

**Review by joeyparrish — APPROVED** [review]

**Comment by avelad** [comment]
It seems that your change introduce an error, can you review it?

**Comment by joeyparrish** [comment]
I'm on it

**Pull request merged**